### PR TITLE
FOUNDATIONS: Support for Aggregate Exceptions on fluent assertion and SameExceptionAs overload

### PR DESCRIPTION
**Pull request by cjdutoit**
This change adds support for:

- Fluent assertion now consider all levels of inner exceptions and not just the 1st two,
- Fluent assertion now also consider `InnerExceptions` that are specific to aggregate exceptions
- SameExceptionAs now has an overload method that will give you the error details.   This is very handy to spot object differences on mock Verify checks

To make use of this you can make the following change in the test:
```cs
            this.loggingBrokerMock.Verify(broker =>
                broker.LogError(It.Is(
                    SameExceptionAs(expectedSourceValidationException, "loggingBrokerMock.Verify", output))),
                        Times.Once);
```

Modify the `SameExceptionAs` in the main test file to now call a new method `IsSameExceptionAs`  This method will do the same comparison, but make use of the SameExceptionAs overload so that we can get the error details.  If any error details are found, it will now be visible in the Test Explorer windows as part of the test result.
```cs
        private static Expression<Func<Xeption, bool>> SameExceptionAs(
            Xeption expectedException, string prefix, ITestOutputHelper output) =>
                actualException => IsSameExceptionAs(actualException, expectedException, prefix, output);

        private static bool IsSameExceptionAs(
            Xeption actualException,
            Xeption expectedException,
            string prefix,
            ITestOutputHelper output)
        {
            var result = actualException.SameExceptionAs(expectedException, out string message);

            if (!string.IsNullOrWhiteSpace(message))
            {
                if (!string.IsNullOrEmpty(prefix))
                {
                    output.WriteLine($"{prefix}: {message}");
                }
                else
                {
                    output.WriteLine(message);
                }
            }

            return result;
        }
```
image

### CODE REVIEW

File: Xeption.Tests/XeptionAssertionTests.cs [hunk 1]
```diff
@@ -3,15 +3,38 @@
...
     {
+        private static int GetRandomNumber() =>
+            new IntRange(min: 2, max: 10).GetValue();
+
         private static string GetRandomString() =>
             new MnemonicString().GetValue();
 
+        private static KeyValuePair<string, List<string>> GenerateKeyValuePair(int count, string keyName = "")
+        {
+            if (string.IsNullOrWhiteSpace(keyName))
```
Comment (hassanhabib) [inline Xeption.Tests/XeptionAssertionTests.cs:22]: Use `String` upper case for class not primitive type

File: Xeption.Tests/XeptionAssertionTests.cs [hunk 2]
```diff
@@ -3,15 +3,38 @@
...
+            {
+                keyName = GetRandomString();
+            }
+
+            List<string> values = Enumerable.Range(start: 0, count: count)
```
Comment (hassanhabib) [inline Xeption.Tests/XeptionAssertionTests.cs:27]: I don't think you have to put the parameter name for `count` but the `start` is legit.

File: Xeption.Tests/XeptionTests.Logic.DataEqualsWithMessage.cs [hunk 3]
```diff
@@ -118,17 +126,21 @@ public void ShouldReturnFalseAndMessageStringIfExpectedDataContainsKeysMatchingK
             string actualValues = ((List<string>)actualXeption.Data[randomKey])
                 .Select(value => value).Aggregate((t1, t2) => t1 + "','" + t2);
 
+            StringBuilder expectedMessage = new StringBuilder();
```
Comment (hassanhabib) [inline Xeption.Tests/XeptionTests.Logic.DataEqualsWithMessage.cs:129]: `var` on the left

File: Xeption/Assertions/XeptionAssertions.cs [hunk 4]
```diff
@@ -25,66 +27,241 @@ public XeptionAssertions(TException exception) : base(exception)
...
-                    "exception data to match but found: "
-                    + Environment.NewLine + exceptionDataComparisonResult.Message)
-                .Then
-                .ForCondition(subject => innerExceptionDataComparisonResult.IsEqual == true)
-                .FailWith(
-                    "inner exception data to match but found: "
-                    + Environment.NewLine + innerExceptionDataComparisonResult.Message)
-                .Then
-                .ClearExpectation();
+            if (Subject is AggregateException actualAggregate && expectation is AggregateException expectedAggregate)
```
Comment (hassanhabib) [inline Xeption/Assertions/XeptionAssertions.cs:30]: I think you can do `and` instead of `&&` here

File: Xeption/Assertions/XeptionAssertions.cs [hunk 5]
```diff
@@ -25,66 +27,241 @@ public XeptionAssertions(TException exception) : base(exception)
...
         }
 
+        private bool InnerExceptionsMatch(
+            Exception actual,
+            Exception expected,
+            string type,
+            string because,
+            params object[] becauseArgs)
+        {
+            if (actual == null && expected == null)
```
Comment (hassanhabib) [inline Xeption/Assertions/XeptionAssertions.cs:69]: Use `is null` - we use arithmatic operators for arithmatic operations only

File: Xeption/Assertions/XeptionAssertions.cs [hunk 6]
```diff
@@ -25,66 +27,241 @@ public XeptionAssertions(TException exception) : base(exception)
...
+                Execute.Assertion
+                    .BecauseOf(because, becauseArgs)
+                    .FailWith(
+                        $"Expected {type} exception to be \"{expected?.GetType().FullName ?? "null"}\", " +
+                        $"but found \"{actual?.GetType().FullName ?? "null"}\".");
+
+                return false;
+            }
+
+            var typeMatch = actual.GetType().FullName == expected.GetType().FullName;
```
Comment (hassanhabib) [inline Xeption/Assertions/XeptionAssertions.cs:83]: What's the type of `typeMatch` here?

File: Xeption/Assertions/XeptionAssertions.cs [hunk 7]
```diff
@@ -25,66 +27,241 @@ public XeptionAssertions(TException exception) : base(exception)
...
+            var messageMatch = actual.Message == expected.Message;
```
Comment (hassanhabib) [inline Xeption/Assertions/XeptionAssertions.cs:84]: What's the type here?

File: Xeption/XeptionExtensions.cs [hunk 8]
```diff
@@ -17,15 +18,30 @@ public static class XeptionExtensions
...
+            try
+            {
+                exception.Should().BeEquivalentTo(otherException);
+                message = string.Empty;
+                return true;
+            }
+            catch (Exception error)
+            {
+                message = error.Message;
+                return false;
```
Comment (hassanhabib) [inline Xeption/XeptionExtensions.cs:43]: New line before `return` and do the same for the rest

File: Xeption/Assertions/XeptionAssertions.cs [hunk 9]
```diff
@@ -25,66 +27,241 @@ public XeptionAssertions(TException exception) : base(exception)
...
+                    .BecauseOf(because, becauseArgs)
+                    .FailWith(errorMessage);
+
+                return false;
+            }
+
+            return true;
+        }
+
+
```
Comment (glhays) [inline Xeption/Assertions/XeptionAssertions.cs:264]: Remove extra line break.

File: Xeption/Assertions/XeptionAssertions.cs (hunk 4 above)
Comment (cjdutoit) [reply Xeption/Assertions/XeptionAssertions.cs:30]: image
Don't have support for that as we are targeting netstandard2.0

File: Xeption/Assertions/XeptionAssertions.cs (hunk 6 above)
Comment (cjdutoit) [reply Xeption/Assertions/XeptionAssertions.cs:83]: Type is compared here to make sure its the same exception.  Then typeMatch is used insude the if and else for the evaluation of: `.ForCondition(typeMatch)` and `.ForCondition(messageMatch)`

File: Xeption/XeptionExtensions.cs (hunk 8 above)
Comment (cjdutoit) [reply Xeption/XeptionExtensions.cs:44]: fixed

File: Xeption.Tests/XeptionAssertionTests.cs (hunk 2 above)
Comment (cjdutoit) [reply Xeption.Tests/XeptionAssertionTests.cs]: fixed

File: Xeption.Tests/XeptionTests.Logic.DataEqualsWithMessage.cs (hunk 3 above)
Comment (cjdutoit) [reply Xeption.Tests/XeptionTests.Logic.DataEqualsWithMessage.cs]: fixed

File: Xeption.Tests/XeptionAssertionTests.cs (hunk 1 above)
Comment (cjdutoit) [reply Xeption.Tests/XeptionAssertionTests.cs]: fixed

File: Xeption/Assertions/XeptionAssertions.cs (hunk 7 above)
Comment (cjdutoit) [reply Xeption/Assertions/XeptionAssertions.cs:84]: It compares the full namespace of the actual and expected exceptions i.e.  `Xeptions.Xeption` is not the same as `STX.Xeption`

File: Xeption/Assertions/XeptionAssertions.cs (hunk 9 above)
Comment (cjdutoit) [reply Xeption/Assertions/XeptionAssertions.cs]: fixed

File: Xeption/Assertions/XeptionAssertions.cs (hunk 7 above)
Comment (glhays) [reply Xeption/Assertions/XeptionAssertions.cs:84]: @cjdutoit I think Hassan means be explicit with the type instead of var.

Comment (cjdutoit) [reply Xeption/Assertions/XeptionAssertions.cs]: fixed

File: Xeption/Assertions/XeptionAssertions.cs [hunk 10]
```diff
@@ -25,66 +27,240 @@ public XeptionAssertions(TException exception) : base(exception)
...
+                            actualInnerExceptions.Count);
+
+                    return false;
+                }
+
+                if (countMatch)
+                {
+                    for (int i = 0; i < actualInnerExceptions.Count; i++)
+                    {
+                        if (!InnerExceptionsMatch(
```
Comment (hassanhabib) [inline Xeption/Assertions/XeptionAssertions.cs:135]: use `is not` instead of `!`

Comment (cjdutoit) [reply Xeption/Assertions/XeptionAssertions.cs]: fixed

**Pull request merged**